dali/operators/decoder/nvjpeg/nvjpeg2k_helper.h

```diff
@@ -23,29 +23,85 @@
-    nvjpeg2kStatus_t status = code;                     \
-    if (status != NVJPEG2K_STATUS_SUCCESS) {            \
-      auto error = make_string("NVJPEG2K error \"",     \
-        static_cast<int>(status), "\"");                \
-      DALI_FAIL(error);                                 \
-    }                                                   \
-  } while (0)
-
-#define NVJPEG2K_CALL_EX(code, extra)                   \
-  do {                                                  \
-    nvjpeg2kStatus_t status = code;                     \
-    std::string extra_info = extra;                     \
-    if (status != NVJPEG2K_STATUS_SUCCESS) {            \
-      auto error = make_string("NVJPEG2K error \"",     \
-        static_cast<int>(status), "\" : ", extra_info); \
-      DALI_FAIL(error);                                 \
-    }                                                   \
-  } while (0)
+class Nvjpeg2kError : public std::runtime_error {
+ public:
+  explicit Nvjpeg2kError(nvjpeg2kStatus_t result, const char *details = nullptr)
+  : std::runtime_error(Message(result, details))
+  , result_(result) {}
+
+  static const char *ErrorString(nvjpeg2kStatus_t result) {
+    switch (result) {
+      case NVJPEG2K_STATUS_SUCCESS:
+        return "The API call has finished successfully. Note that many of the calls are "
+                "asynchronous and some of the errors may be seen only after synchronization.";
+      case NVJPEG2K_STATUS_NOT_INITIALIZED :
+        return "The library handle was not initialized.";
+      case NVJPEG2K_STATUS_INVALID_PARAMETER:
+        return "Wrong parameter was passed. For example, a null pointer as input data, or "
+               "an invalid enum value";
+      case NVJPEG2K_STATUS_BAD_JPEG:
+        return "Cannot parse the JPEG2000 stream. Likely due to a corruption that cannot "
+               "be handled";
+      case NVJPEG2K_STATUS_JPEG_NOT_SUPPORTED:
+        return "Attempting to decode a JPEG2000 stream that is not supported by "
+               "the nvJPEG2000 library.";
+      case NVJPEG2K_STATUS_ALLOCATOR_FAILURE:
+        return "The user-provided allocator functions, for either memory allocation or "
+               "for releasing the memory, returned a non-zero code.";
+      case NVJPEG2K_STATUS_EXECUTION_FAILED:
+        return "Error during the execution of the device tasks.";
+      case NVJPEG2K_STATUS_ARCH_MISMATCH:
+        return "The device capabilities are not enough for the set of input parameters provided.";
+      case NVJPEG2K_STATUS_INTERNAL_ERROR:
+        return "Unknown error occurred in the library.";
+      case NVJPEG2K_STATUS_IMPLEMENTATION_NOT_SUPPORTED:
+        return "API is not supported by the backend.";
+      default:
+        return "< unknown error >";
+    }
+  }
+
+  static std::string Message(nvjpeg2kStatus_t result, const char *details) {
+    if (details && *details) {
+      return make_string("nvjpeg2k error: ", result, " ", ErrorString(result),
```

Comment:
```suggestion
      return make_string("nvJPEG2000 error ", result, ": ", ErrorString(result),
```
The official name of the library is *nvJPEG2000*.
Also, I think that this looks better:
`nvJPEG2000 error 2: Wrong parameter was passed.`
than that:
`nvJPEG2000 error, 2, Wrong parameter was passed.`

Reply:
Done